golang/session/channel.go

```diff
@@ -0,0 +1,251 @@
+package session
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/progrium/qmux/golang/codec"
+)
+
+type channelDirection uint8
+
+const (
+	channelInbound channelDirection = iota
+	channelOutbound
+)
+
+// channel is an implementation of the Channel interface that works
+// with the session class.
+type channel struct {
+	ctx context.Context
```

Comment:
I guess the ctx doesn't do anything at all yet?

Reply:
There were two cases I had in mind. 

The first was timeouts. More APIs that have a long running, cancelable bit of work usually take a context for timeouts, but what are reasonable timeouts here? Maybe a session/conn can be canceled? If `net.Conn` doesn't care then I guess we wouldn't here. However, a Channel could have a context because a Channel is not necessarily a connection, it could be a request. So looking at how net/http uses them with request is probably helpful. 

The second was contextual meta-data. Context becomes a a key/value store that is passed down a thread of work that can contain relevant data or objects that come from up the calling code. There is much less consensus on the right way to use this, as you can argue those should be part of the API. However, too many times in my experience it can be used as a way to pass data in that the author didn't think to allow, so it's just a helpful thing to have. But I think that's also important to pass data that is not relevant to the domain of the library, but a domain it might be used. If you use Channel as a request, maybe you want to keep a trace ID with it. 

So I took the approach of being able to add a context to a session, which was passed to a channel, so there is a hierarchy to top level values that might be associated with a session you can have with a channel if you are passing just a channel around. Maybe the channel should have its own context in which the session is in. You could say maybe that should be part of the API, but I think it makes more sense in a context.

So yea.

Reply:
After studying net and http packages and reading the article, I've removed contexts since there are other ways to associate values with a channel (wrapping it in a Request that may have its own context setup). The only context that makes sense is for Open, which I've run into situations where I'd like it to timeout, so I added that.